### ISSUE TITLE
Support DWARF in wasm-bindgen

### ISSUE THREAD

**Issue by mitsuhiko** [rustwasm/wasm-bindgen#2389]
### Motivation

Chrome, Firefox and now also Sentry support DWARF for WebAssembly as debug format. In particularly for crash reporting systems like Sentry DWARF is a very important tool to create accurate stack traces for production builds of web assembly applications.

This issue was previously brought up in #1981.

### Proposed Solution

Most likely the bulk of the implementation will be in walrus (see https://github.com/rustwasm/walrus/issues/67).

**Comment by Cypher1** [comment]
Just hoping for an update as this would be really handy for debugging WASM performance issues. Thx

**Comment by michael8090** [comment]
After some experiments, I find it's possible to show Rust code in the Chrome DevTools with the following steps:

1. add a `--keep-debug` in `wasm-bindgen` when building the wasm file, i.e. `wasm-bindgen --keep-debug --out-dir ...`
2. install a C++ chrome extension and enable DWARF support feature in chrome dev tools: https://developer.chrome.com/blog/wasm-debugging-2020/?utm_source=devtools

Then the rust code can show up in the dev tools, but other functionalities such as setting breakpoints and evaluating rust code, are broken:

<img width="1182" alt="wecom-temp-bcb7d09421dfc8d9b0d40550932bdf05" src="https://user-images.githubusercontent.com/2306105/158933271-a8171d22-0829-4dfb-ba28-037df59d3cfa.png">


It seems the DWARF info is kept when using wasm-bindgen with `--keep-debug`, but chrome don't fully understand the dwarf info for rust code.

**Comment by CryZe** [comment]
It seems more like the debug info is "broken" after wasm-bindgen processes it, as in enough to make the source show up, but not enough for Chrome to handle it.

**Comment by michael8090** [comment]
> It seems more like the debug info is "broken" after wasm-bindgen processes it, as in enough to make the source show up, but not enough for Chrome to handle it.

The extension is for C++ only, and you can see when I try to evaluate `ElementsManager::get_static()` in console, it takes the string as C++ code and says it's an invalid one. My guess is that maybe the dwarf info is correct, but it's the C++ extension that cannot understand rust code. I hope maybe someone inside the extension project or wasm-bindgen can give some concrete information about it.

**Comment by CryZe** [comment]
Yeah, calling methods and stuff almost definitely won't work, but local variables and their fields, and breakpoints are all things that should work and last I tested DO work... if you don't use wasm-bindgen. So that still seems like a strong indication that the dwarf is broken.

**Comment by Saruniks** [comment]
Maybe this issue will provide you more info about the possible reasons why it doesn't work it wasm-bindgen https://github.com/rustwasm/walrus/issues/67

**Comment by stan-irl** [comment]
@CryZe did breakpoints work for you?? The devtools find my rust code but none of the standard library or cargo packages. They are all prefixed like `/cargo/...` or `/rustc/...`

**Comment by RReverser** [comment]
> but none of the standard library or cargo packages. They are all prefixed like `/cargo/...` or `/rustc/...`

It's expected and happens because standard library in Rust is prebuilt on CI with those paths hardcoded. You can remap them to either github or your local rust-src paths via the extension settings.

But, either way, the breakpoints won't work at the moment because, as mentioned above, DWARF remapping is not implemented in wasm-bindgen.

**Comment by stan-irl** [comment]
got it! thankyou for the quick response 🙏

**Comment by RReverser** [comment]
> But, either way, the breakpoints won't work at the moment because, as mentioned above, DWARF remapping is not implemented in wasm-bindgen.

Should probably clarify that this is only if you use wasm-bindgen. If you use raw Wasm, it will work (but that's uncommon on the web).

**Comment by stan-irl** [comment]
ah i got. because wasm-bindgen is actually changing the wasm when it generates the bindings but its not changing the DWARF so nothing matches up. 

Regular wasm doesnt do that so its not a problem

**Comment by Cldfire** [comment]
> Since DWARF doesn't seem to work well with Chrome (at least on Windows), we'd also need make our own version of [`wasm-sourcemap.py`](https://github.com/emscripten-core/emscripten/blob/main/tools/wasm-sourcemap.py) (since a Python dependency on top of Rust and JS would be a bit much).

Someone has now built this, and playing around with it locally it seems like it works well: https://github.com/mtolmacs/wasm2map

**Comment by RReverser** [comment]
@Cldfire FWIW that comment above was mistaken, DWARF does work well with Chrome directly: https://github.com/rustwasm/walrus/pull/231#issuecomment-1431619795

**Comment by d3lm** [comment]
Yea, I can also confirm that DWARF works well for me with Chrome, at least on macOS (M1).

**Comment by connor4312** [comment]
VS Code's JavaScript debugger is also now DWARF aware (nightly build today, and to stable in 2 weeks). Support here would make targeting wasm from Rust much more pleasant for VS/Code folks 🙂

**Comment by daxpedda** [comment]
Just for visibility: this is currently only blocked on https://github.com/rustwasm/walrus/issues/247.

**Comment by eiswind** [comment]
for visibility: https://github.com/rustwasm/walrus/issues/247 has been fixed.

**Comment by connor4312** [comment]
Awesome! Can't wait to give this a shot on Monday... so I'll probably try it this weekend 😄

**Comment by squillace** [comment]
Well..... @connor4312 ????? I'm on tenterhooks and I believe this is more important than Thanksgiving because if it works _I can give thanks!_

**Comment by haraldreingruber-dedalus** [comment]
How is it decided when new wasm-bindgen versions are released? On demand?
Looking forward to giving the next release a try 😎

**Comment by connor4312** [comment]
>  I'm on tenterhooks and I believe this is more important than Thanksgiving because if it works I can give thanks!

DWARF symbols seems to work with a few caveats:

- Publishing of packages still needs to happen. Also, I can only get symbols working if I invoke wasm-bindgen directly, not via the more common `wasm-pack`, but maybe this'll be fixed when things are published?
- Locations work in the Chrome C++ extension but they're off in VS Code's debugger. This is likely on my side to fix; we [consume the guts](https://github.com/microsoft/vscode-dwarf-debugging) of the Chrome plugin to do the DWARF heavy lifting so I'm probably just holding it wrong. I will look after the holiday.

**Comment by d3lm** [comment]
> How is it decided when new wasm-bindgen versions are released? On demand?

@haraldreingruber-dedalus I think this should be released. Or maybe I am mistaken.

**Comment by connor4312** [comment]
Fixed in VS Code, will be released in about 1.5 weeks from today

<img width="627" alt="image" src="https://github.com/rustwasm/wasm-bindgen/assets/2230985/d978227e-be66-4335-b4ad-7f50aeb9cc3c">

**Comment by RReverser** [comment]
@connor4312 It looks like the call stack name is not taken from DWARF but rather from the names section? Is that a currently known limitation?

**Comment by connor4312** [comment]
that's formed from asking the worker for `getFunctionInfo`, which is the same ([from what I understand](https://github.com/ChromeDevTools/devtools-frontend/blob/c9f204731633fd2e2b6999a2543e99b7cc489b4b/docs/language_extension_api.md#dealing-with-inlined-functions)) as what the devtools extension does. I get the same result there, maybe the issue is upstream?

<img width="324" alt="image" src="https://github.com/rustwasm/wasm-bindgen/assets/2230985/2ccfc329-3b10-43c9-abeb-0e33b978efe4">

**Comment by RReverser** [comment]
Ah weird, could be. It's been couple of years since I tried Rust-produced Wasm in DevTools. I guess they'd need Rust-specific demangling to omit it. Not a big deal either way, thanks for working on this!

**Comment by brentmjohnson** [comment]
@connor4312 thanks so much for the work on this! I kind of expected this to be released to vscode-insiders but maybe not yet? rustwasm/walrus dependency upgrades are now in the 0.2.89 release ([diff](https://github.com/rustwasm/wasm-bindgen/compare/0.2.88...0.2.89#diff-d43310c69dbd64dbe201466da69d7a4f3bd4662bbed8dd7fbab668709553627e))

![image](https://github.com/rustwasm/wasm-bindgen/assets/84997972/b423c22f-e49e-4db4-9740-69d1c22e2ee9)

**Comment by connor4312** [comment]
It's in tomorrow's Insiders

**Comment by michael8090** [comment]
Thanks for the wonderful work. I played for a while and found the following:

1. breakpoint, stepping: works nicely, great
2. local variable inspecting: it shows `undefined` for every local variable, no idea if it's something I did wrong:
<img width="1165" alt="image" src="https://github.com/rustwasm/wasm-bindgen/assets/2306105/ef42302d-fdf0-4038-aff7-ba0ecc8b380e">
3. function arguments inspecting: arguments are `undefined` except `self`. `self` can be fully expanded, and some `owned` values can be inspected properly, but boxed values are hidden behind pointers, which makes them hard to be inspected: 
<img width="1064" alt="image" src="https://github.com/rustwasm/wasm-bindgen/assets/2306105/860087b3-fab3-4c3d-92b3-8f60f516981a">
